feat: update KycOnboarding to allow for top ups, update Manager to allow for config-only updates

nickmzero opened the pull request:
#### KycOnboarding.sol changes:
- Adds canTopUp configuration. 
- New environment variable: `KYC_CAN_TOP_UP`, 0 or 1
- Coupon message updated to include memberNonce. Now using memberNonces mapping (address => uint256) instead of redeemedCoupons to prevent replays.

#### Manager.sol changes:
- UpdateType.CONFIGS added to allow for config only updates.

fforbeck commented on tasks/deploy.js:
```diff
@@ -650,6 +659,7 @@ const deployAvalancheDao = async ({
       getEnvVar("DAO_OWNER_ADDR")
     ),
     kycMaxMembers: getOptionalEnvVar("KYC_MAX_MEMBERS", toBN(1000)),
+    kycCanTopUp: getOptionalEnvVar("KYC_CAN_TOP_UP", 0),
```
I am wondering if we should leave the top up enabled by default `1`. I think it makes sense to double check that, otherwise it just adds some overhead to update every DAO after the deployment to enable the top up.

nickmzero replied:
this makes sense to me. Are there any docs that should be updated to reflect this change in behavior? @adridadou thoughts?

fforbeck commented on contracts/adapters/KycOnboarding.sol:
```diff
@@ -80,9 +83,8 @@ contract KycOnboardingContract is
-    // hashCouponMessage(dao, Coupon(kycedMember)) => bool
-    mapping(bytes32 => bool) public redeemedCoupons;
+    // memberAddress => uint256
+    mapping(address => uint256) public memberNonces;
```
Good idea.

fforbeck commented on test/adapters/kyc-onboarding.test.js:
```diff
@@ -418,11 +454,131 @@ describe("Adapter - KYC Onboarding", () => {
     ).to.be.revertedWith("already redeemed");
   });
+
+  it("should be possible to first join when configured with canTopUp", async () => {
```
We probably need a test case where the member joins for the first time and the top up is disabled. It is related to the use case I mentioned to you on the `KycOnboarding.sol` review.

fforbeck commented on test/adapters/kyc-onboarding.test.js:
```diff
@@ -418,11 +454,131 @@ describe("Adapter - KYC Onboarding", () => {
+    expect(applicantUnits.toString()).equal(
+      numberOfUnits.mul(toBN("6")).toString()
+    );
+  });
```
I didn't see that but we could add a test case to cover the coupon replay. WDYT?

nickmzero replied:
yes I'll add

fforbeck reviewed (changes requested):
It looks great @nickmzero !
I added some comments and questions about the top up and the new update type.

nickmzero commented on contracts/adapters/KycOnboarding.sol:
```diff
@@ -243,22 +264,28 @@ contract KycOnboardingContract is
         require(
-            !isActiveMember(dao, dao.getCurrentDelegateKey(kycedMember)),
+            !isActiveMember(dao, dao.getCurrentDelegateKey(kycedMember)) ||
+                _daoCanTopUp(dao, tokenAddr),
```
In that case it's technically not a top up, but the first redemption and top ups share the same logic.
`!isActiveMember || _daoCanTopUp` and `!isActiveMember || (isActiveMember && _daoCanTopUp)` should resolve the same way. Do we need a distinction on whether an onboarding is a top up or not?

fforbeck replied:
Gotcha. `!isActiveMember || _daoCanTopUp` should be enough. The first time the `kycMember` is not active, so that will pass. Any operation after that, for the same `kycMember`, will be verified with the `_daoCanTopUp`. It is all good then. Regarding the operation type, I don't think need that distinction here at this stage, but it would be nice to double check with @adridadou. One idea would be to change the `Onboarded` event to support a `topUp` flag.

adridadou replied:
I don't think that knowing whether it was a top up or not is important information. If we have previous events then it is a top up. 

The only thing I see here that we need to take into account is that the change in EIP-712 structure means that we need to change how the kyc attestation is created on the kyc backend side to add the nonce.

nickmzero commented on contracts/adapters/Manager.sol:
```diff
@@ -160,7 +161,10 @@ contract Manager is Reimbursable, AdapterGuard, Signatures {
         dao.submitProposal(proposalId);
         dao.processProposal(proposalId);
-        if (proposal.updateType == UpdateType.ADAPTER) {
+        if (proposal.updateType == UpdateType.CONFIGS) {
```
makes sense, I'll update this

nickmzero commented:
> It looks great @nickmzero ! I added some comments and questions about the top up and the new update type.

Pushed these changes:
- allowing extension acl updates with CONFIGS flag
- added test for extension acl updates with flag
- added kyc coupon replay test
- added kyc first join with canTopUp disabled test
- changed KYC_CAN_TOP_UP to default to 1, will reach out to David to double check we're fine with this

fforbeck reviewed (dismissed):
Let's double check if we will need to change the `topUp` default config, and the `Onboarded` event. Other than that it looks great to me @nickmzero !
